File: src/liveScoring.ts
export type LiveScoreEntry = {
  rank: number;
  rider_name: string;
  horse_name: string;
  finishing_score: number;
  dressage_score: number;
  show_jumping_penalties: number;
  cross_country_jump_penalties: number;
  cross_country_time_penalties: number;
  source_id: string;
  source_record_id: string;
  collected_at: string;
};

export type LiveScoreEvent = {
  event_name: string;
  event_date: string;
  level: string;
  country: string;
  latest_collected_at: string;
  entry_count: number;
  leader: LiveScoreEntry | null;
  entries: LiveScoreEntry[];
};

export type LiveScoreFeed = {
  version: number;
  source_id: string;
  updated_at: string;
  window_start: string | null;
  window_end: string | null;
  event_count: number;
  score_count: number;
  events: LiveScoreEvent[];
};

const LIVE_FEED_URL = '/live_scores.json';

export const loadLiveScoreFeed = async (url = LIVE_FEED_URL): Promise<LiveScoreFeed | null> => {
  if (typeof fetch === 'undefined') {
    return null;
  }

  try {
    const response = await fetch(url, { cache: 'no-store' });
    if (!response.ok) {
      return null;
    }

    const payload: unknown = await response.json();
    return isLiveScoreFeed(payload) ? payload : null;
  } catch {
    return null;
  }
};

export const formatFeedDateRange = (feed: Pick<LiveScoreFeed, 'window_start' | 'window_end'>) => {
  if (!feed.window_start && !feed.window_end) {
    return 'current event window';
  }

  if (feed.window_start === feed.window_end) {
    return feed.window_start ?? 'current event window';
  }

  return `${feed.window_start ?? 'open'} to ${feed.window_end ?? 'open'}`;
};

export const formatTimestamp = (value: string) => {
  const timestamp = new Date(value);
  if (Number.isNaN(timestamp.getTime())) {
    return value;
  }

  return new Intl.DateTimeFormat(undefined, {
    dateStyle: 'medium',
    timeStyle: 'short',
  }).format(timestamp);
};

const isLiveScoreFeed = (value: unknown): value is LiveScoreFeed => {
  if (!isRecord(value)) {
    return false;
  }

  return (
    value.version === 1 &&
    typeof value.source_id === 'string' &&
    typeof value.updated_at === 'string' &&
    typeof value.event_count === 'number' &&
    typeof value.score_count === 'number' &&
    Array.isArray(value.events) &&
    value.events.every(isLiveScoreEvent)
  );
};

const isLiveScoreEvent = (value: unknown): value is LiveScoreEvent => {
  if (!isRecord(value)) {
    return false;
  }

  return (
    typeof value.event_name === 'string' &&
    typeof value.event_date === 'string' &&
    typeof value.level === 'string' &&
    typeof value.country === 'string' &&
    typeof value.latest_collected_at === 'string' &&
    typeof value.entry_count === 'number' &&
    (value.leader === null || isLiveScoreEntry(value.leader)) &&
    Array.isArray(value.entries) &&
    value.entries.every(isLiveScoreEntry)
  );
};

const isLiveScoreEntry = (value: unknown): value is LiveScoreEntry => {
  if (!isRecord(value)) {
    return false;
  }

  return (
    typeof value.rank === 'number' &&
    typeof value.rider_name === 'string' &&
    typeof value.horse_name === 'string' &&
    typeof value.finishing_score === 'number' &&
    typeof value.dressage_score === 'number' &&
    typeof value.show_jumping_penalties === 'number' &&
    typeof value.cross_country_jump_penalties === 'number' &&
    typeof value.cross_country_time_penalties === 'number' &&
    typeof value.source_id === 'string' &&
    typeof value.source_record_id === 'string' &&
    typeof value.collected_at === 'string'
  );
};

const isRecord = (value: unknown): value is Record<string, unknown> =>
  typeof value === 'object' && value !== null;

File: equibets/live_scoring.py
"""Live score feed helpers for current eventing results."""

from __future__ import annotations

import json
from collections.abc import Iterable, Sequence
from datetime import date, datetime, timedelta, timezone
from pathlib import Path

from equibets.results import EventingResult, consolidate_results


DEFAULT_LIVE_SCORE_FILE = Path(__file__).resolve().parents[1] / "public" / "live_scores.json"


def current_event_window(
    today: date | None = None,
    *,
    lookback_days: int = 2,
    lookahead_days: int = 3,
) -> tuple[date, date]:
    """Return the date range to refresh for events that may still be changing."""

    if lookback_days < 0 or lookahead_days < 0:
        raise ValueError("lookback_days and lookahead_days must be non-negative")

    anchor = today or datetime.now(timezone.utc).date()
    return anchor - timedelta(days=lookback_days), anchor + timedelta(days=lookahead_days)


def build_live_score_feed(
    results: Iterable[EventingResult],
    *,
    window_start: date | None = None,
    window_end: date | None = None,
    updated_at: datetime | None = None,
    source_id: str = "data_fei",
) -> dict[str, object]:
    """Build a JSON-serializable live scoring feed grouped by event and level."""

    consolidated = [
        result
        for result in consolidate_results(list(results))
        if _within_window(result.event_date, window_start, window_end)
    ]
    scoreboards = _scoreboards(consolidated)
    timestamp = (updated_at or datetime.now(timezone.utc)).replace(microsecond=0)

    return {
        "version": 1,
        "source_id": source_id,
        "updated_at": timestamp.isoformat(),
        "window_start": window_start.isoformat() if window_start else None,
        "window_end": window_end.isoformat() if window_end else None,
        "event_count": len(scoreboards),
        "score_count": len(consolidated),
        "events": scoreboards,
    }


def save_live_score_feed(
    results: Sequence[EventingResult],
    path: Path | str = DEFAULT_LIVE_SCORE_FILE,
    *,
    window_start: date | None = None,
    window_end: date | None = None,
    updated_at: datetime | None = None,
    source_id: str = "data_fei",
) -> dict[str, object]:
    """Write the live score feed and return the payload that was saved."""

    payload = build_live_score_feed(
        results,
        window_start=window_start,
        window_end=window_end,
        updated_at=updated_at,
        source_id=source_id,
    )
    output_path = Path(path)
    output_path.parent.mkdir(parents=True, exist_ok=True)
    with output_path.open("w", encoding="utf-8") as feed_file:
        json.dump(payload, feed_file, indent=2, sort_keys=True)
        feed_file.write("\n")
    return payload


def _scoreboards(results: Sequence[EventingResult]) -> list[dict[str, object]]:
    grouped: dict[tuple[str, date, str], list[EventingResult]] = {}
    for result in results:
        grouped.setdefault((result.event_name, result.event_date, result.level), []).append(result)

    scoreboards: list[dict[str, object]] = []
    for (event_name, event_date, level), event_results in grouped.items():
        ranked_results = sorted(
            event_results,
            key=lambda result: (
                result.finishing_score,
                result.rider_name.lower(),
                result.horse_name.lower(),
            ),
        )
        entries = [_entry_payload(result, rank) for rank, result in enumerate(ranked_results, start=1)]
        leader = entries[0] if entries else None
        country = _first_known_country(ranked_results)
        latest_collected_at = max(result.collected_at for result in ranked_results)
        scoreboards.append(
            {
                "event_name": event_name,
                "event_date": event_date.isoformat(),
                "level": level,
                "country": country,
                "latest_collected_at": latest_collected_at.isoformat(),
                "entry_count": len(entries),
                "leader": leader,
                "entries": entries,
            }
        )

    return sorted(
        scoreboards,
        key=lambda event: (
            str(event["event_date"]),
            str(event["event_name"]).lower(),
            str(event["level"]).lower(),
        ),
        reverse=True,
    )


def _entry_payload(result: EventingResult, rank: int) -> dict[str, object]:
    return {
        "rank": rank,
        "rider_name": result.rider_name,
        "horse_name": result.horse_name,
        "finishing_score": result.finishing_score,
        "dressage_score": result.dressage_score,
        "show_jumping_penalties": result.show_jumping_penalties,
        "cross_country_jump_penalties": result.cross_country_jump_penalties,
        "cross_country_time_penalties": result.cross_country_time_penalties,
        "source_id": result.source_id,
        "source_record_id": result.source_record_id,
        "collected_at": result.collected_at.isoformat(),
    }


def _within_window(result_date: date, window_start: date | None, window_end: date | None) -> bool:
    if window_start and result_date < window_start:
        return False
    if window_end and result_date > window_end:
        return False
    return True


def _first_known_country(results: Sequence[EventingResult]) -> str:
    for result in results:
        if result.country and result.country != "Unknown":
            return result.country
    return "Unknown"
